addons/stock/wizard/stock_picking_return.py

```diff
@@ -116,6 +116,7 @@ def _prepare_picking_default_values(self):
             'move_ids': [],
             'picking_type_id': self.picking_id.picking_type_id.return_picking_type_id.id or self.picking_id.picking_type_id.id,
             'state': 'draft',
+            'partner_id': self.picking_id.partner_id.id,
```

Comment:
This doesn't replace the logic of `_create_returns` removed, it was specifically in `purchase_stock` to handle the following case:
- set 3 step incoming shipments on the warehouse.
- create and confirm a PO.
- confirm the receipt `Vendor > Input`.
- confirm `Input > Quality`.
- the return the `Input >  Quality` transfer.
- choose the Vendor as return location.

The transfer created does not have a contact (partner_id), because there is no `partner_id` on `Input > Quality` transfer, to handle this:
override this method (`_prepare_picking_default_values`) in `purchase_stock` instead and add partner_id to vals like you did here but use the partner_id from the `move_ids` instead,  just like the logic in `_create_returns` you removed, don't set the `partner_id` in vals if there is more than one `move_ids.partner_id`.